### DISCUSSION TITLE
Possible to get 0 compression pixel perfect native resolution without rescaling?

### DISCUSSION THREAD

**Discussion by ZeChArtiahSaher** [apohl79/audiogridder#1037] [Help]
I briefly looked at code , there is some rescaling seems to be going on that I noticed in `Screen.cpp`. Is there a faster way to have 1:1 image send? As I'm assuming that either rescaling or maybe compression somewhere in the pipeline makes the image lose sharpness both in ffmpeg - high & legacy quality 1.

**Comment by apohl79** [comment]
Everything is going through a compression format (webp for ffmpeg and png for legacy). If you need higher quality, the compression parameters for webp could be made manually adjustable. But non compressed images are not supported.

**Reply by apohl79** [reply]
Looks like webp has a lossless mode. Will have a look at what parameters it needs for this, so it might solve for what you are looking for.

**Reply by ZeChArtiahSaher** [reply]
Yeah, ic , it's avcodec everywhere, so maybe with the right config to &opts to libwebp might bear some results, unless there is extra comp or whatever to take care of, I'll have a look as well, there should be lossless option like you said

**Reply by ZeChArtiahSaher** [reply]
So, tried making some changes to see whether simply getting lossless would help.

https://github.com/ZeChArtiahSaher/audiogridder/pull/1/files

* Was testing osx plugin & win host
* With lossless weird color conversion problems occur (at least for windows host gridder server grigrab (BRGA) & webp) somehow - Ended up working around via ARGB (server output codec pix fmt) -> BGRA (plugin input codec pix fmt)
* If plugin host has scale > 1, mby option to preserve scale of the window? 
* Even after trying to match dpis and bilinear, seems like there is still something in the way (maybe just viewport sizing artifacts?): __(left is plugin, right is server host)__
![img](https://i.imgur.com/DXakYlK.png)

**Reply by apohl79** [reply]
Yeah I see it. Maybe the problem is that the capture rect is not necessarily the same as the window rect on the desktop because of this:

`avcodec_align_dimensions(m_outputCodecCtx, &m_outputCodecCtx->width, &m_outputCodecCtx->height);`

I believe I added this some time back to fix some random memory corruption errors.

**Reply by apohl79** [reply]
Maybe try different scaling algorithms here:

```
    m_swsCtx = sws_getContext(m_captureRect.getWidth(), m_captureRect.getHeight(), m_captureCodecCtx->pix_fmt,
                              m_outputFrame->width, m_outputFrame->height, m_outputCodecCtx->pix_fmt, SWS_BICUBIC,
                              nullptr, nullptr, nullptr);
```
see https://ffmpeg.org/doxygen/2.7/swscale_8h_source.html (line 55-66). Maybe that gives better results?

**Reply by ZeChArtiahSaher** [reply]
Yeah, removing align dimensions is almost there.

I wonder why u were gettin mem leaks though, POIs could maybe incl. post assumptions based on implicit conversion (yuv), something like linesize mby? (Wild guess). I read somewhere in the docs that linesize is usually mod 32. Maybe irrelevant though.

https://github.com/apohl79/audiogridder/blob/master/Server/Source/ScreenRecorder.cpp#L353

In terms of sws scaling billinear should be most transparent for 1:1/downscale. Of course, won't be best for upscaling. (Could technically switch dynamically based on down vs up).
 https://www.programmerall.com/article/27511952318/

I poked around a bit, also had to adjust mouse pos according to plugin host scaling, thoe that seems to introduce slight error margin. Should compare to master beforehand.

**Reply by apohl79** [reply]
> I wonder why u were gettin mem leaks though

no leaks, random crashes, that went away after using aligned buffers.

**Reply by apohl79** [reply]
Just taking out the `avcodec_align_dimensions` call removes all the blur on my end. So thats the reason. It might be needed to check the buffer allocation. There seems to be some conditions where it crashed the scaler I believe. Need to see if I still find some of those.

**Reply by ZeChArtiahSaher** [reply]
Alright, btw, is there any streamlined debugging you use? Link is enough. Currently I rely on building and checkin logs which not ideal and dev docs don't got much.

Also for some reason couldn't get loggin to work on the server in the record loop. Taking out:

https://github.com/apohl79/audiogridder/blob/c61a59c04bda7b64ae298b82397729d9d4f5b891/Server/Source/ScreenRecorder.cpp#L35

Does make logs work in `initialize`, but doesn't help with (even with I use something like `setLogTagStatic()`:

https://github.com/apohl79/audiogridder/blob/c61a59c04bda7b64ae298b82397729d9d4f5b891/Server/Source/ScreenRecorder.cpp#L246-L247

Obviously, probably I should be doing this differently completely, I did notice JUCE_DEBUG, but not sure where to specify exactly and obv what is the debugging method you've got.

Also some easy VST debugging methods possibly? Sorry to ask so many questions at once :D

**Reply by apohl79** [reply]
> Alright, btw, is there any streamlined debugging you use? Link is enough. Currently I rely on building and checkin logs which not ideal and dev docs don't got much.

Depends on what I'm looking for. Most of the time I run the server and/or DAW in lldb, set break points etc. Sometimes I'm using the tracer. If you create a debug build, there is a target called tracereader, that you can use to read trace files. This is sometimes needed to understand what was going on in the different treads prior to a crash or to debug things that can't be printed to the log.

Once you run the server or plugins in lldb (or gdb on linux or the visual studio debugger on windows), all logging goes directly to the console (as well as to the log files). You can also turn on logging to the console outside of the debugger (for the server only) by passing `-log` as a command line param.

> Also for some reason couldn't get loggin to work on the server in the record loop. 

Funny... just added ffmpeg logging, so that errors end up in the normal AG log. Pushed the changes, so you can pull again. 

I also removed the width/height alignment on the output context, but I added it to calculate the output frame buf size to get it a bit bigger, just in case. Will monitor sentry with the next beta closely to see, if the scaler crashes are coming back. Everything is stable here.

In general, all logging is implemented in the Logger.hpp/.cpp files. There is a dedicated log thread.

> Also some easy VST debugging methods possibly? Sorry to ask so many questions at once :D

See above. Run the DAW in a debugger.
